packages/waku/src/router/inferred-router-types.ts

```diff
@@ -0,0 +1,91 @@
+import type { AnyPage, PathWithoutSlug } from './create-pages.js';
+import type { Join, ReplaceAll, Split } from './util-types.js';
+
+type StaticSlugPage = {
+  path: string;
+  render: 'static';
+  staticPaths: (string | string[])[];
+};
+
+type DynamicPage = {
+  path: string;
+  render: 'dynamic';
+};
+
+type IsPageWithSlug<Page extends AnyPage> = Page extends {
+  path: infer P;
+  render: 'static' | 'dynamic';
+}
+  ? P extends PathWithoutSlug<P>
+    ? false
+    : true
+  : never;
+
+/** Used to replace a single slug param [mySlug] with a union of possible slugs 'foo' | 'bar' */
+type ReplaceSlugSet<
+  Path extends string,
+  Slugs extends string,
+> = Slugs extends unknown ? ReplaceAll<Path, `[${string}]`, Slugs> : never;
+
+type ReplaceHelper<
+  SplitPath extends string[],
+  Slugs extends string[],
+  // SlugCountArr is a counter for the number of slugs added to result so far
+  SlugCountArr extends null[] = [],
+  Result extends string[] = [],
+> = SplitPath extends [infer PathPart, ...infer Rest]
+  ? Rest extends string[]
+    ? PathPart extends string
+      ? PathPart extends `[...${string}]`
+        ? [...Result, ...Slugs] // Wildcard always comes last
+        : PathPart extends `[${string}]`
+          ? ReplaceHelper<
+              Rest,
+              Slugs,
+              [...SlugCountArr, null],
+              [...Result, Slugs[SlugCountArr['length']]]
+            >
+          : ReplaceHelper<Rest, Slugs, SlugCountArr, [...Result, PathPart]>
+      : never
+    : never
+  : Result;
+
+type ReplaceTupleStaticPaths<
+  Path extends string,
+  StaticPathSet extends string[],
+> = StaticPathSet extends unknown
+  ? Join<ReplaceHelper<Split<Path, '/'>, StaticPathSet>, '/'>
+  : never;
+
+type CollectPathsForStaticSlugPage<Page extends StaticSlugPage> = Page extends {
+  path: infer Path;
+  render: 'static';
+  staticPaths: infer StaticPaths;
+}
+  ? Path extends string
+    ? StaticPaths extends string[]
+      ? ReplaceSlugSet<Path, StaticPaths[number]>
+      : StaticPaths extends string[][]
+        ? ReplaceTupleStaticPaths<Path, StaticPaths[number]>
+        : never
+    : never
+  : never;
+
+type CollectPathsForDynamicSlugPage<Page extends DynamicPage> = Page extends {
+  path: infer Path;
+}
+  ? Path extends string
+    ? ReplaceAll<Path, `[${string}]`, string>
+    : never
+  : never;
+
+export type CollectPaths<EachPage extends AnyPage | void> = // void is returned by createLayout
+  EachPage extends AnyPage
+    ? IsPageWithSlug<EachPage> extends true
+      ? EachPage extends StaticSlugPage
+        ? CollectPathsForStaticSlugPage<EachPage>
+        : EachPage extends DynamicPage
+          ? CollectPathsForDynamicSlugPage<EachPage>
+          : never
+      : EachPage['path']
+    : never;
```

Comment:
Hmm, in the source code, I would like to make it clear that:
- "router" is something we define in `unstable_defineRouter`.
- "createPages" is **one instance** of "router".
- Assume there can be another instance of router, like `createFoo` independent from `createPages`.

So, any code specific to `create-pages.ts` should probably move to `src/router/create-pages-utils/*.ts`

Reply:
actually `src/router/create-pages/*.ts` is fine, if you prefer. My request is to avoid `index.ts`.

Reply:
Ok, how's this look?

```
packages/waku/src/router
├── base-types.ts
├── client.ts
├── common.ts
├── create-pages
│   ├── common.ts
│   ├── inferred-router-types.ts
│   └── types-from-route-config.ts
├── define-router.ts
├── fs-router.ts
├── server.ts
└── util-types.ts

2 directories, 10 files
```

Reply:
> * "router" is something we define in `unstable_defineRouter`.

For this, would it be good enough to add jsdoc comments to the types in the inferred-router-types.ts to explain where each is called from and what they're used for?

> * "createPages" is **one instance** of "router".
> * Assume there can be another instance of router, like `createFoo` independent from `createPages`.

I was thinking that when we support other router types that these types would still be used to infer the final paths for consistency on our end. For example `AnyRouter` would be widened to support other "router"'s and then from that point on, all the types for path inference would be the same.

Reply:
yeah, jsdoc would be helpful.

We don't know the future and how `createFoo` looks like, if it shares some thing, we can restructure. For now, I would like to separate unstable_defineRouter and createPages very clearly. For example, slugs are only for createPages.

Reply:
added jsdoc to inferred-path-types ✅